lib/katello/tasks/update_subscription_facet_backend_data.rake

```diff
@@ -0,0 +1,61 @@
+namespace :katello do
+  task :update_subscription_facet_backend_data => ["environment"] do
+    def error(message)
+      @errors << message
+      @errors << "\n"
+    end
+
+    def report_errors
+      if @errors.any?
+        filename = "subscription_facet_upgrade-#{Time.now.to_i}.log"
+        path = "/var/log/foreman/#{filename}"
+        path = "/tmp/#{filename}" unless File.writable?(path)
+
+        file = File.open(path, 'w')
+        @errors.each { |error| file.write(error) }
+        file.close
+        $stderr.print "***********************************\n"
+        $stderr.print "*************WARNING***************\n"
+        $stderr.print "Errors detected during upgrade step.\n"
+        $stderr.print "Details saved to: #{file.path}\n"
+        $stderr.print "This step can be rerun with:\n"
+        $stderr.print "  foreman-rake katello:update_subscription_facet_backend_data\n"
+        $stderr.print "You are likely encountering a bug.\n"
+        $stderr.print "***********************************\n"
+      end
+    end
+
+    @errors ||= []
+    User.current = User.anonymous_api_admin
+    puts _("Updating backend data for subscription facets")
+
+    #there may be some invalid hosts, if there are create a primary interface
+    ::Host.includes(:interfaces).find_each do |host|
+      if host.primary_interface.nil?
+        host.interfaces.create!(:primary => true)
```

Comment:
I think the discussion diverged a bit here. What @jlsherrill said in second comment is true for any invalid host but the original concern was for hosts without the primary interface which I think is not easy to cause and generally should not happen. I'm not sure how often will this task be run, probably just once during the upgrade so I'm OK with keeping the hack there. But if it gets copied for each new task I think the code will be full of hacks no one understands :-)